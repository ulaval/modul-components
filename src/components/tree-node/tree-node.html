<li class="m-tree-node" ref="item">
    <m-link :mode="linkMode" class="m-tree-node__single-node" :class="{ 'm--is-selected': isSelected }" :disabled="linkDisabled"
        :underline="false" @click="onClick" ref="link">
        <m-tree-icon class="m-tree-icon" v-if="icons" :filename="node.id" :folder-open="open" :folder="isFolder"></m-tree-icon>
        <span>{{ label }}</span>
    </m-link>
    <div v-if="isFolder" class="m-tree-node__with-children">
        <m-accordion-transition v-if="hasChildren">
            <ul v-show="open" class="m-tree-node__with-children__body" :class="{'m-tree-node__line-extension' : hasSibling}" ref="childNodeList">
                <m-tree-node v-for="(childNode, index) in node.children" :key="childNode.id" class="m-tree-node__with-children__children"
                    :path="currentPath" :icons="icons" :selectable="selectable" :node="childNode" :selected-nodes="selectedNodes"
                    :has-sibling="childHasSibling(index)" @click="onChildClick"></m-tree-node>
            </ul>
        </m-accordion-transition>
        <m-accordion-transition v-else>
            <ul v-show="open" class="m-tree-node__with-children__body" :class="{'m-tree-node__line-extension' : hasSibling}">
                <li class="m-tree-node__with-children__children--empty">
                    {{emptyContentMessage}}
                </li>
            </ul>
        </m-accordion-transition>
    </div>
</li>
