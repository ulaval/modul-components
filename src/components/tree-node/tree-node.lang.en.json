{
    "m-tree-node": {
        "empty": "No content"
    }
}
