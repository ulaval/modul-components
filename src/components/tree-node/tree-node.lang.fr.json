{
    "m-tree-node": {
        "empty": "Aucun contenu"
    }
}
