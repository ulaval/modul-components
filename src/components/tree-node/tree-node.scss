@import '../../styles/abstracts/variables';
@import '../accordion/accordion-transition';

$m-tree-node--height: $m-touch-size;
$m-tree-node--child-indentation: 52px;
$m-tree-node--line-width: 22px;

.m-tree-node {
    color: $m-color--ul-blue;

    & + &,
    &__children > & {
        margin-top: $m-margin;
    }

    &__item {
        border-radius: $m-tree-node--height / 2;
        padding-left: 21px;
        padding-right: 21px;
        width: $m-min-width;
        display: flex;
        align-items: center;
        height: $m-tree-node--height;
        @include m-button();
        @include m-button--secondary();

        &__label {
            white-space: nowrap;
            overflow: hidden;
            text-overflow: ellipsis;
            line-height: $m-line-height;
        }
    }

    &.m--is-disabled > &__item {
        @include m-button--disabled();
    }

    &.m--is-selected > &__item,
    &:not(.m--is-disabled) > &__item:hover {
        @include m-button--hover();
    }

    &__line-extension::before {
        content: "";
        position: absolute;
        left: -$m-tree-node--line-width;
        width: 1px;
        top: 0;
        bottom: 0;
        background-color: $m-color--interactive;
    }

    &__child {
        margin-left: $m-tree-node--child-indentation;
        position: relative;

        &::before {
            content: "";
            position: absolute;
            left: -$m-tree-node--line-width;
            top: $m-tree-node--height / 2 - 1;
            width: $m-tree-node--line-width;
            height: 1px;
            background-color: $m-color--interactive;
        }

        &.m--is-disabled::before {
            background-color: $m-color--disabled;
        }

        &::after {
            content: "";
            position: absolute;
            left: -$m-tree-node--line-width;
            width: 1px;
            top: -$m-margin;
            height: $m-tree-node--height + $m-margin;
            background-color: $m-color--interactive;
        }

        &:last-child::after {
            height: $m-tree-node--height / 2 + $m-margin;
        }

        &-empty {
            margin-left: $m-tree-node--child-indentation - $m-tree-node--line-width;
            margin-top: $m-margin;
            color: $m-color--text;
        }
    }
}
