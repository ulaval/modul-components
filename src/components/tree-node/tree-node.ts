import Vue, { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';

import { ModulVue } from '../../utils/vue/vue';
import { TREE_NODE_NAME } from '../component-names';
import I18nPlugin from '../i18n/i18n';
import IconPlugin from '../icon/icon';
import TreeIconPlugin from '../tree-icon/tree-icon';
import { TreeNode } from '../tree/tree';
import WithRender from './tree-node.html?style=./tree-node.scss';

@WithRender
@Component
export class MTreeNode extends ModulVue {
    @Prop()
    public node: TreeNode;

    @Prop({ default: false })
    public open: boolean;

    @Prop({ default: [] })
    public selectedNodes: string[];

    @Prop()
    selectable: boolean;

    @Prop()
    public icons: boolean;

    @Prop({ default: '' })
    public path: string;

    @Prop()
    public placeholder: string;

    @Prop()
    public hasSibling: boolean;

    public internalOpen: boolean = false;

    public onClick(): void {
        if (this.isFolder) {
            this.internalOpen = !this.internalOpen;
            this.$emit('update:open', this.internalOpen);
        } else if (this.selectable) {
            this.$emit('click', this.currentPath);
        }
    }

    public onChildClick(path: string): void {
        this.$emit('click', path);
    }

    public childHasSibling(index: number): boolean {
        if (this.node.children) {
            return !!this.node.children[index + 1];
        }
        return false;
    }

    protected mounted(): void {
        this.internalOpen = this.open ? this.open : this.isParentOfSelectedFile;
    }

    public get currentPath(): string {
        return this.path + '/' + this.node.id;
    }

    public get label(): string {
        return this.node.label || this.node.id;
    }

    private get isParentOfSelectedFile(): boolean {
        let pathMatchesSelectedNode: boolean = false;
        this.selectedNodes.forEach(selectedNode => {
            let reg: RegExp = new RegExp(`${this.currentPath}\/`);
            pathMatchesSelectedNode = reg.test(selectedNode) || pathMatchesSelectedNode;
        });
        return pathMatchesSelectedNode;
    }

    public get hasChildren(): boolean {
        return !!this.node.children && this.node.children.length > 0;
    }

    public get isFolder(): boolean {
        return this.node.hasChildren || !!this.node.children;
    }

    public get isDisabled(): boolean {
        return !this.selectable && !this.isFolder;
    }

    public get isSelected(): boolean {
        return this.selectedNodes.indexOf(this.currentPath) !== -1;
    }

    public get emptyContentMessage(): string {
        return this.placeholder || this.$i18n.translate('m-tree-node:empty');
    }
}

const TreeNodePlugin: PluginObject<any> = {
    install(v, options): void {
        v.prototype.$log.debug(TREE_NODE_NAME, 'plugin.install');
        v.use(TreeIconPlugin);
        v.use(IconPlugin);
        v.use(I18nPlugin);
        v.component(TREE_NODE_NAME, MTreeNode);
    }
};

export default TreeNodePlugin;
