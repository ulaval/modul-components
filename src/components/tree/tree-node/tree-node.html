<li class="m-tree-node"
    :class="{ 'm--is-selected': isSelected && !isDisabled,
            'm--is-disabled' : isDisabled,
            'm--is-read-only' : isReadonlyStyle,
            'm-tree-node__has-children' : isFolder,
            'm-tree-node__with-checkboxes' : withCheckboxes,
            'm-tree-node__file-icons' : useFilesIcons }">
    <div class="m-tree-node__item"
         ref="item"
         role="button"
         :class="{ 'm-tree-node__item--no-hover': moveSelectionZoneToCheckbox }"
         :tabindex="isDisabled ? '-1' : 0"
         @click="onClick">
        <m-tree-icon v-if="useFilesIcons"
                     class="m-tree-icon"
                     :is-folder-open="internalOpen"
                     :is-folder="isFolder"></m-tree-icon>
        <m-tree-icon class="m-tree-icon"
                     :filename="node.id"
                     :is-folder-open="internalOpen"
                     :use-files-icons="useFilesIcons"
                     :is-folder="isFolder"></m-tree-icon>
        <m-checkbox v-if="withCheckboxes"
                    ref="checkbox"
                    :value="isSelected"
                    :indeterminate="isIndeterminate"
                    :disabled="isDisabled"
                    :readonly="readonly"
                    @change="onCheckboxClick"></m-checkbox>
        <span class="m-tree-node__item-label">{{label}}</span>
        <span class="m-tree-node__item-check"></span>
        <m-icon v-if="hasRightIcon"
                class="m-tree-node__warning-icon"
                :name="node.rightIconName"
                size="1em"></m-icon>
    </div>
    <template v-if="isFolder">
        <m-accordion-transition v-if="hasChildren">
            <ul class="m-tree-node__children"
                ref="children"
                v-show="internalOpen"
                :class="{'m-tree-node__line-extension' : hasSibling}">
                <m-tree-node class="m-tree-node__child"
                             v-for="(childNode, index) in node.children"
                             :key="childNode.id"
                             :path="currentPath"
                             :selectable="selectable"
                             :selected="isSelected"
                             :readonly="readonly"
                             :open.sync="childNode.open"
                             :node="childNode"
                             :use-files-icons="useFilesIcons"
                             :selected-nodes="selectedNodes"
                             :disabled-nodes="disabledNodes"
                             :checkboxes="checkboxes"
                             :has-sibling="childHasSibling(index)"
                             @auto-select-child-checkbox-change="onAutoSelectChildCheckboxChange"
                             @click="onChildClick"></m-tree-node>
            </ul>
        </m-accordion-transition>
        <m-accordion-transition v-else>
            <ul class="m-tree-node__children"
                v-show="internalOpen"
                :class="{'m-tree-node__line-extension' : hasSibling}">
                <li class="m-tree-node__child-empty">
                    {{emptyContentMessage}}
                </li>
            </ul>
        </m-accordion-transition>
    </template>
    <template v-if="displaySelectionButton">
        <button class="m-tree-node__auto-select-button"
                ref="autoSelectButton"
                :class="{ 'm-tree-node__auto-select-button--selected': allChildrenAndSelfSelected}"
                @click="onAutoSelectButtonClick"></button>
    </template>
</li>
