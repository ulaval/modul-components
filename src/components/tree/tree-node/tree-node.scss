@import '../../../styles/abstracts/variables';
@import '../../accordion/accordion-transition';

$m-tree-node--height: $m-touch-size;
$m-tree-node--child-indentation: 52px;
$m-tree-node--line-width: 22px;

.m-tree-node {
    color: $m-color--ul-blue;
    margin-left: $m-tree-node--line-width;
    position: relative;

    &::before {
        content: "";
        position: absolute;
        left: -$m-tree-node--line-width;
        top: $m-tree-node--height / 2 - 1;
        width: $m-tree-node--line-width;
        height: 1px;
        background-color: $m-color--interactive;
    }

    &.m--is-disabled::before {
        background-color: $m-color--disabled;
    }

    &::after {
        content: "";
        position: absolute;
        left: -$m-tree-node--line-width;
        width: 1px;
        top: $m-tree-node--height / 2;
        bottom: -$m-tree-node--height / 2 - $m-margin;
        background-color: $m-color--interactive;
    }

    &:last-child::after {
        content: none;
    }

    & + &,
    &__children > & {
        margin-top: $m-margin;
    }

    &__auto-select-button {
        position: absolute;
        left: $m-min-width + 10px;
        top: $m-tree-node--height / 4;
        height: $m-tree-node--height / 2;
        background: red;

        &--selected {
            background: blue;
        }
    }

    &__item {
        border-radius: $m-tree-node--height / 2;
        padding-left: 21px;
        padding-right: 21px;
        width: $m-min-width;
        display: flex;
        align-items: center;
        height: $m-tree-node--height;

        @include m-button();
        @include m-button--secondary();

        &__label {
            white-space: nowrap;
            overflow: hidden;
            text-overflow: ellipsis;
            line-height: $m-line-height;
            flex-grow: 1;
        }

        .m-tree-node__item__label:first-child {
            .m-tree--multiple-node-checkbox & {
                margin-left: $m-spacing--m;
            }
        }

        /deep/ .m-checkbox__input {
            margin-left: 8px;
        }
    }

    &.m--is-disabled > &__item {
        @include m-button--disabled();
    }

    &.m--is-selected > &__item {
        @include m-button--active();
    }

    @media (min-width: $m-mq--min--s) {
        &:not(.m--is-disabled):not(.m--is-selected) > &__item:not(.m-tree-node__item--no-hover):hover {
            @include m-button--hover();
        }

        &:not(.m--is-disabled) > &__item:hover,
        &.m--is-selected {
            .m-tree-icon::before,
            .m-tree-icon::after {
                background-color: $m-color--white;
            }
        }
    }

    &__line-extension::before {
        content: "";
        position: absolute;
        left: -$m-tree-node--line-width;
        width: 1px;
        top: 0;
        bottom: 0;
        background-color: $m-color--interactive;
    }

    &__child {
        margin-left: $m-tree-node--child-indentation;

        &::after {
            top: -$m-margin;
            height: $m-tree-node--height + $m-margin;
        }

        &:last-child::after {
            content: "";
            height: $m-tree-node--height / 2 + $m-margin;
        }

        &-empty {
            margin-left: $m-tree-node--child-indentation - $m-tree-node--line-width;
            margin-top: $m-margin;
            color: $m-color--text;
        }
    }
}

.m-tree__content > .m-tree-node:only-child {
    margin-left: 0;

    &::before {
        content: none;
    }
}
