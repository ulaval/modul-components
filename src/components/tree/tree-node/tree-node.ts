import Component from 'vue-class-component';
import { Emit, Prop, Watch } from 'vue-property-decorator';
import { ModulVue } from '../../../utils/vue/vue';
import { TREE_ICON_NAME } from '../component-names';
import { MCheckboxes, TreeNode } from '../tree';
import { MTreeIcon } from '../tree-icon/tree-icon';
import WithRender from './tree-node.html?style=./tree-node.scss';

@WithRender
@Component({
    components: {
        [TREE_ICON_NAME]: MTreeIcon
    }
})
export class MTreeNode extends ModulVue {
    @Prop()
    public node: TreeNode;

    @Prop({ default: false })
    public open: boolean;

    @Prop({ default: [] })
    public selectedNodes: string[];

    @Prop()
    public checkboxes: MCheckboxes;

    @Prop()
    public selectable: boolean;

    @Prop()
    public useFilesIcons: boolean;

    @Prop({ default: '' })
    public path: string;

    @Prop()
    public placeholder: string;

    @Prop()
    public hasSibling: boolean;

    @Prop()
    public disabledNodes: string[];

    public internalOpen: boolean = false;

    public allChildrenAndSelfSelected: boolean = false;

    private selectedChildrenCount: number = 0;

    @Watch('isSelected')
    public notifyParentOfChildCheckboxState(): void {
        if (this.withCheckboxes) {
            if (!this.hasChildren) {
                this.$emit('auto-select-child-checkbox-change', this.isSelected);
            } else if (this.isButtonAutoSelect && this.hasChildren) {
                this.onAutoSelectChildCheckboxChange(this.isSelected, true);
            }
        }
    }

    public onClick(): void {
        if (this.isFolder) {
            this.internalOpen = !this.internalOpen;
            this.$emit('update:open', this.internalOpen);
        } else if (this.selectable && !this.withCheckboxes) { // Clicks are emitted from checkboxes when true
            this.$emit('click', this.currentPath);
        }
    }

    @Emit('click')
    public onChildClick(path: string): string {
        return path;
    }

    public onAutoSelectButtonClick(): void {
        this.recursiveSelect();
    }

    public onAutoSelectChildCheckboxChange(selected: boolean, ignoreCount: boolean = false): void {
        if (!ignoreCount) {
            this.selectedChildrenCount += selected ? 1 : -1;
        }
        let allChildrenSelected: boolean = this.selectedChildrenCount === (this.node.children ? this.node.children.length : -1);

        if (this.isCheckboxAutoSelect || (this.isParentAutoSelect && !selected)) { // Parent checkbox mode does not select parent by itself
            this.updateCheckboxParentNode(allChildrenSelected);
        } else if (this.isButtonAutoSelect) {
            this.updateButtonParentNode(allChildrenSelected, selected);
        }
    }

    public childHasSibling(index: number): boolean {
        if (this.node.children) {
            return !!this.node.children[index + 1];
        }
        return false;
    }

    public onCheckboxClick(): void {
        if (this.hasChildren && (this.isCheckboxAutoSelect || this.isParentAutoSelect)) {
            this.recursiveSelect();
        } else {
            this.$emit('click', this.currentPath);
        }
    }

    protected mounted(): void {
        this.internalOpen = this.open ? this.open : this.isParentOfSelectedFile;
        if (this.isSelected) {
            this.notifyParentOfChildCheckboxState();
        }
    }

    private recursiveSelect(): void {
        let childrenPaths: string[] = [];
        let addNodesToSelected: boolean = false;

        if (this.isButtonAutoSelect) {
            childrenPaths.concat(this.currentPath);
            if (!this.allChildrenAndSelfSelected) {
                addNodesToSelected = true;
            }
        } else if (!this.isButtonAutoSelect && !this.isSelected) {
            addNodesToSelected = true;
        }

        this.fetchChildrenPaths(this.node, childrenPaths, this.path + '/' + this.node.id, (this.isParentAutoSelect || this.isButtonAutoSelect));
        this.updateSelectedNodes(childrenPaths, addNodesToSelected);
    }

    private updateSelectedNodes(paths: string[] = [], addNodesToSelected: boolean): void {
        paths = Array.from(new Set(paths));
        paths.forEach(path => {
            let nodeAlreadySelected: boolean = this.selectedNodes.indexOf(path) !== -1;
            if ((addNodesToSelected && !nodeAlreadySelected) || (!addNodesToSelected && nodeAlreadySelected)) { // Prevent nodes to be pushed twice or unselected nodes to be removed
                this.$emit('click', path);
            }
        });
    }

    private fetchChildrenPaths(currentNode: TreeNode, childrenPath: string[], path: string, includeParent: boolean = false): void {
        if (currentNode.children && currentNode.children.length > 0) {
            currentNode.children.forEach(child => {
                this.fetchChildrenPaths(child, childrenPath, path + '/' + child.id, includeParent);
            });
            if (includeParent) {
                childrenPath.push(path);
            }
        } else {
            childrenPath.push(path);
        }
    }

    private updateCheckboxParentNode(allChildrenSelected: boolean): void {
        if (allChildrenSelected && !this.isSelected) {
            this.$emit('click', this.currentPath); // Auto-push current, emit to parent for recursivity
            this.$emit('auto-select-child-checkbox-change', true);
        } else if (this.isSelected) {
            this.$emit('click', this.currentPath);
            this.$emit('auto-select-child-checkbox-change', false); // Not all children are selected, stop recursivity
        }
    }

    private updateButtonParentNode(allChildrenSelected: boolean, selected: boolean): void {
        if (allChildrenSelected && selected && this.isSelected) { // New element selected was added, current and all children are selected
            this.allChildrenAndSelfSelected = true;
            this.$emit('auto-select-child-checkbox-change', true); // Notifies immidiate parent's button that current button is on
        } else if (this.allChildrenAndSelfSelected) { // Button was on, but something was removed.
            this.allChildrenAndSelfSelected = false;
            this.$emit('auto-select-child-checkbox-change', false); // Something was removed so every parent buttons must know
        }
    }

    public get isParentOfSelectedFile(): boolean {
        let pathMatchesSelectedNode: boolean = false;
        this.selectedNodes.forEach(selectedNode => {
            let parentPath: string = this.currentPath + '/';
            if (selectedNode.indexOf(parentPath) !== -1) {
                pathMatchesSelectedNode = true;
            }
        });
        return pathMatchesSelectedNode;
    }

    public get isDisabled(): boolean {
        let isDisabled: boolean = false;
        let inDisabledNodes: boolean = this.disabledNodes && this.disabledNodes.indexOf(this.currentPath) !== -1;
        if (!this.selectable && !this.isFolder || inDisabledNodes) {
            isDisabled = true;
        }
        return isDisabled;
    }

    public get currentPath(): string {
        return this.path + '/' + this.node.id;
    }

    public get label(): string {
        return this.node.label || this.node.id;
    }

    public get propCheckboxes(): MCheckboxes {
        return this.checkboxes || MCheckboxes.False;
    }

    public get moveSelectionZoneToCheckbox(): boolean {
        return this.withCheckboxes && !this.isFolder;
    }

    public get displaySelectionButton(): boolean {
        return this.hasChildren && this.checkboxes === MCheckboxes.WithButtonAutoSelect;
    }

    public get hasChildren(): boolean {
        return !!this.node.children && this.node.children.length > 0;
    }

    public get isFolder(): boolean {
        return this.node.hasChildren || !!this.node.children;
    }

    public get isSelected(): boolean {
        return this.selectedNodes.indexOf(this.currentPath) !== -1;
    }

    public get isIndeterminate(): boolean {
        return (this.isCheckboxAutoSelect || this.isParentAutoSelect) && this.isParentOfSelectedFile && !this.isSelected;
    }

    public get emptyContentMessage(): string {
        return this.placeholder || this.$i18n.translate('m-tree-node:empty');
    }

    public get withCheckboxes(): boolean {
        return this.checkboxes !== MCheckboxes.False;
    }

    private get isParentAutoSelect(): boolean {
        return this.checkboxes === MCheckboxes.WithParentAutoSelect;
    }

    private get isButtonAutoSelect(): boolean {
        return this.checkboxes === MCheckboxes.WithButtonAutoSelect;
    }

    private get isCheckboxAutoSelect(): boolean {
        return this.checkboxes === MCheckboxes.WithCheckboxAutoSelect;
    }
}
