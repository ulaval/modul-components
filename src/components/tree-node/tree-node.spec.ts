import { RefSelector, shallow, Wrapper } from '@vue/test-utils';

import { renderComponent } from '../../../tests/helpers/render';
import { TreeNode } from '../tree/tree';
import { MTreeNode } from './tree-node';

const CHILDREN: RefSelector = { ref: 'children' };
const ITEM: RefSelector = { ref: 'item' };

const NODE_ELEMENT_LABEL: string = 'Node 1';
const NODE_ELEMENT_ID: string = 'Node 1';
const PARENT_PATH: string = '/Parent 1';
const SELECTION_ICON: string = 'information';

const TREE_NODE_SELECTED: string[] = ['/Node 1'];
const TREE_NODE_SELECTED_2: string[] = ['/Node 2'];
const TREE_NODE_SELECTED_3: string[] = ['/Node 3/Node 4'];

const TREE_NODE_WITHOUT_CHILDREN: TreeNode = {
    label: NODE_ELEMENT_LABEL,
    id: NODE_ELEMENT_ID
};

const TREE_NODE_WITHOUT_CHILDREN_NOT_VALID: TreeNode = {
    label: NODE_ELEMENT_LABEL,
    id: '',
    children: []
};

const TREE_NODE_WITH_CHILDREN_EMPTY: TreeNode = {
    label: 'Node 2',
    id: 'Node 2',
    hasChildren: true,
    children: []
};

const TREE_NODE_WITH_CHILDREN: TreeNode = {
    label: 'Node 3',
    id: 'Node 3',
    children: [
        {
            label: 'Node 4',
            id: 'Node 4',
            children: []
        }
    ]
};

const TREE_NODE_WITH_CHILDREN_NOT_VALID: TreeNode = {
    label: 'Node 3',
    id: '',
    hasChildren: true,
    children: [
        {
            label: 'Node 4',
            id: 'Node 4',
            children: []
        }
    ]
};

let node: TreeNode;
let selectedNodes: string[] = [];
let selectable: boolean = true;
let icons: boolean = false;
let path: string = '';

let wrapper: Wrapper<MTreeNode>;

const initializeShallowWrapper: any = () => {
    wrapper = shallow(MTreeNode, {
        stubs: getStubs(),
        propsData: {
            node,
            selectedNodes,
            selectable,
            icons,
            path
        }
    });
};

const getStubs: any = () => {
    return {
        ['m-tree-icon']: '<div>m-tree-icon</div>',
        ['m-link']: '<a @click="$emit(\'click\')"><slot /></a>'
    };
};

describe('MTreeNode', () => {

    describe(`Given a node`, () => {

        describe(`When the node can't have children`, () => {

            describe(`and the node is valid`, () => {
                beforeEach(() => {
                    node = TREE_NODE_WITHOUT_CHILDREN;
                    initializeShallowWrapper();
                });

                it(`Should render correctly`, () => {
                    expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
                });

                it(`The node label should be "Node 1"`, () => {
                    expect(wrapper.vm.label).toEqual(NODE_ELEMENT_LABEL);
                });

                it(`The node does not have children`, () => {
                    expect(wrapper.vm.hasChildren).toBeFalsy();
                });
            });

            describe(`and the node is not valid`, () => {
                beforeEach(() => {
                    node = TREE_NODE_WITHOUT_CHILDREN_NOT_VALID;
                    initializeShallowWrapper();
                    wrapper.setMethods({ generateErrorTree: jest.fn() });
                });
            });

        });

        describe(`When the node can have children but don't have any`, () => {

            beforeEach(() => {
                node = TREE_NODE_WITH_CHILDREN_EMPTY;
                initializeShallowWrapper();
            });

            it(`Should render correctly`, () => {
                expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
            });

        });

        describe(`When the node can and has children`, () => {

            describe(`and the node is valid`, () => {

                beforeEach(() => {
                    node = TREE_NODE_WITH_CHILDREN;
                    initializeShallowWrapper();
                });

                it(`Should render correctly`, () => {
                    expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
                });

                it(`Should have children`, () => {
                    expect(wrapper.vm.hasChildren).toBeTruthy();
                });

            });

            describe(`When the node is a parent of the selected node`, () => {
                it(`The node should be open`, () => {
                    selectedNodes = TREE_NODE_SELECTED_3;
                    initializeShallowWrapper();

                    expect(wrapper.find(CHILDREN)).toBeTruthy();
                });
            });

            describe(`When the node is not a parent of the selected node`, () => {
                it(`The node should not be open`, () => {

                    node = TREE_NODE_WITH_CHILDREN;
                    selectedNodes = TREE_NODE_SELECTED_2;
                    initializeShallowWrapper();

                    expect(wrapper.find(CHILDREN)).toBeTruthy();
                });
            });

        });

        describe(`When selectionMode = single`, () => {

            beforeEach(() => {
                node = TREE_NODE_WITHOUT_CHILDREN;
                initializeShallowWrapper();
            });

            it(`Should emit "click"`, () => {
                wrapper.find(ITEM).trigger('click');

                expect(wrapper.emitted('click')).toBeTruthy();
            });

        });

        describe(`When not selectable`, () => {

            beforeEach(() => {
                selectable = false;
                node = TREE_NODE_WITHOUT_CHILDREN;
                initializeShallowWrapper();
            });

            it(`Should not be able to select a node`, () => {
                wrapper.find(ITEM).trigger('click');

                expect(wrapper.emitted('click')).toBeFalsy();
            });
        });

        describe(`When we check if the current node is selected`, () => {
            beforeEach(() => {
                node = TREE_NODE_WITHOUT_CHILDREN;
            });

            describe(`When the current node is selected`, () => {

                beforeEach(() => {
                    selectedNodes = TREE_NODE_SELECTED;
                });

                describe(`and selectionIcon is not empty`, () => {
                    beforeEach(() => {
                        initializeShallowWrapper();
                    });

                    it(`The node should be selected`, () => {
                        expect(wrapper.vm.isSelected).toBeTruthy();
                    });
                });
            });

            describe(`When an other node is selected`, () => {
                it(`The node should not be selected`, () => {
                    selectedNodes = TREE_NODE_SELECTED_2;
                    initializeShallowWrapper();

                    expect(wrapper.vm.isSelected).toBeFalsy();
                });
            });

            describe(`When no node is selected`, () => {
                it(`The node should not be selected`, () => {
                    selectedNodes = [];
                    initializeShallowWrapper();

                    expect(wrapper.vm.isSelected).toBeFalsy();
                });
            });
        });

        describe(`When the node has a parent`, () => {

            it(`Should return the right current path`, () => {
                path = PARENT_PATH;
                node = TREE_NODE_WITHOUT_CHILDREN;
                initializeShallowWrapper();

                expect(wrapper.vm.currentPath).toEqual(PARENT_PATH + '/' + NODE_ELEMENT_ID);
            });
        });

    });

});
