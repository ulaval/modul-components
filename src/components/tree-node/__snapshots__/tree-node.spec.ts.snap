// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MTreeNode Given a node When the node can and has children and the node is valid Should render correctly 1`] = `
<li class="m-tree-node">
    <a mode="button" class="m-tree-node__single-node"> <span>Node 3</span></a>
    <div class="m-tree-node__with-children">
        <m-accordion-transition>
            <ul class="m-tree-node__with-children__body" style="display:none;">
                <li class="m-tree-node m-tree-node__with-children__children">
                    <m-link mode="button" class="m-tree-node__single-node"> <span>Node 4</span></m-link>
                    <div class="m-tree-node__with-children">
                        <m-accordion-transition>
                            <ul class="m-tree-node__with-children__body" style="display:none;">
                                <li class="m-tree-node__with-children__children--empty">
                                    m-tree-node:empty
                                </li>
                            </ul>
                        </m-accordion-transition>
                    </div>
                </li>
            </ul>
        </m-accordion-transition>
    </div>
</li>
`;

exports[`MTreeNode Given a node When the node can have children but don't have any Should render correctly 1`] = `
<li class="m-tree-node">
    <a mode="button" class="m-tree-node__single-node"> <span>Node 2</span></a>
    <div class="m-tree-node__with-children">
        <m-accordion-transition>
            <ul class="m-tree-node__with-children__body" style="display:none;">
                <li class="m-tree-node__with-children__children--empty">
                    m-tree-node:empty
                </li>
            </ul>
        </m-accordion-transition>
    </div>
</li>
`;

exports[`MTreeNode Given a node When the node can't have children and the node is valid Should render correctly 1`] = `
<li class="m-tree-node">
    <a mode="button" class="m-tree-node__single-node"> <span>Node 1</span></a>
</li>
`;
